### PR TITLE
Conditionals and operators

### PR DESCRIPTION
**Pull request by agbanusi**

### CODE REVIEW

**Review by rjnrohit — COMMENTED** [review]
Please add behaviour tests for this.

File: src/passes/literalExpressionEvaluator/literalExpressionEvaluator.ts [hunk 1]
```diff
@@ -99,7 +99,7 @@ function evaluateLiteral(node: Literal): RationalLiteral | boolean | null {
 
 function evaluateUnaryLiteral(node: UnaryOperation): RationalLiteral | boolean | null {
   const op = evaluateLiteralExpression(node.vSubExpression);
-  if (op === null) return null;
+  if (!op) return null;
```
Comment (rodrodros) [inline src/passes/literalExpressionEvaluator/literalExpressionEvaluator.ts]: Nitpick: could you revert this change back. I believe that by explicitly stating that you want it to be different than null makes the code safer and more readable

File: src/passes/literalExpressionEvaluator/literalExpressionEvaluator.ts [hunk 2]
```diff
@@ -128,16 +128,15 @@ function evaluateBinaryLiteral(node: BinaryOperation): RationalLiteral | boolean
...
+    const notNullMember = left ?? right;
     if (notNullMember === null) {
       return null;
-    }
-    if (typeof notNullMember === 'boolean') {
+    } else if (typeof notNullMember === 'boolean') {
       switch (node.operator) {
         case '&&': // false && x = false
-          return notNullMember ? null : false;
+          return notNullMember && false;
```
Comment (cicr99) [inline src/passes/literalExpressionEvaluator/literalExpressionEvaluator.ts]: This should be changed back to the way it was before. The way this is right now always returns false, which is unwanted if `notNullMember` is true

File: src/passes/literalExpressionEvaluator/literalExpressionEvaluator.ts [hunk 3]
```diff
@@ -128,16 +128,15 @@ function evaluateBinaryLiteral(node: BinaryOperation): RationalLiteral | boolean
...
         case '||': // true || x = true
-          return notNullMember ? true : false;
+          return notNullMember || true;
```
Comment (cicr99) [inline src/passes/literalExpressionEvaluator/literalExpressionEvaluator.ts]: Same thing here, this always returns true

File: src/passes/literalExpressionEvaluator/literalExpressionEvaluator.ts [hunk 4]
```diff
@@ -236,6 +226,14 @@ function evaluateBinaryLiteral(node: BinaryOperation): RationalLiteral | boolean
...
       case '>>':
         return left.shiftRight(right);
+      case '&':
+        return left.and(right);
+      case '|':
+        return left.or(right);
+      case '^':
+        return left.xor(right);
+      case '~':
+        return left.not();
```
Comment (cicr99) [inline src/passes/literalExpressionEvaluator/literalExpressionEvaluator.ts]: This operator can be removed from this switch because `~` is a unary operator, and this is for binary operators

File: src/passes/literalExpressionEvaluator/rationalLiteral.ts [hunk 5]
```diff
@@ -137,6 +137,28 @@ export class RationalLiteral {
...
+  }
+  xor(other: RationalLiteral): RationalLiteral {
+    return new RationalLiteral(
+      this.numerator ^ other.numerator,
+      this.denominator ^ other.denominator,
+    );
+  }
+  not(): RationalLiteral {
+    return new RationalLiteral(~this.numerator, this.denominator);
+  }
```
Comment (cicr99) [inline src/passes/literalExpressionEvaluator/rationalLiteral.ts]: This functionality is already defined here as `bitwiseNegate`, so no need for this, it can be removed.

It is also a fine idea to add the "bitwise" prefix to these functions as in: `bitwiseAnd`, `bitwiseOr`, `bitwiseXor`; so it can be different from the logical operators

File: tests/behaviour/contracts/expressions/rationalLiterals.sol [hunk 6]
```diff
@@ -52,4 +52,20 @@ contract WARP {
...
+    return 18|6;
+  }
+
+  function xor() pure public returns (int8) {
+    return 18^6;
+  }
+
+  function not() pure public returns (int8) {
+    return ~3;
+  }
```
Comment (cicr99) [inline tests/behaviour/contracts/expressions/rationalLiterals.sol]: This can be removed as well, since there is a case above for it

**Review by cicr99 — CHANGES_REQUESTED** [review]

**Comment by agbanusi** [comment]
> Please add behaviour tests for this.

Done this

**Review by cicr99 — APPROVED** [review]

**Review by piwonskp — APPROVED** [review]

**Review by rodrodros — APPROVED** [review]
